### ISSUE TITLE
Segfault

### ISSUE THREAD

**Issue by wrathematics** [keroro824/HashingDeepLearning#6]
This looks very interesting, but I'm unable to successfully run the example. I modified the `trainData`, `testData`, and `logFile` paths of the config file appropriately. Valgrind reports this:

```
==12259== Memcheck, a memory error detector
==12259== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==12259== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==12259== Command: ./runme Config_amz.csv
==12259== 
new Network
==12259== Invalid write of size 4
==12259==    at 0x10F124: Network::Network(int*, NodeType*, int, int, float, int, int*, int*, int*, float*, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, cnpy::NpyArray, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, cnpy::NpyArray> > >) (Network.cpp:13)
==12259==    by 0x10D01E: main (main.cpp:472)
==12259==  Address 0xb is not stack'd, malloc'd or (recently) free'd
```

**Comment by keroro824** [comment]
Apologize for the late reply! The team just got back from the conference. We'd happy to help. As mentioned in the README: "Additionally, Transparent Huge Pages must be enabled. SLIDE requires approximately 900 2MB pages, and 10 1GB pages. (https://wiki.debian.org/Hugepages)"

May I ask if you that is enabled?

**Comment by wrathematics** [comment]
No worries. I wouldn't call your reply late at all!

I saw the THP requirement in the readme, but I'm not familiar with it so I'm not really sure. I see this:

```
$ cat /sys/kernel/mm/transparent_hugepage/enabled
always [madvise] never
```

Reading `man 2 madvise`, it's not clear to me exactly how this works. But I also get the segfault if I remove usage of `MAP_HUGETLB` and `MAP_HUGE_SHIFT` from Layer.h, Network.h, and Node.h.

**Comment by keroro824** [comment]
In order to omit the possibility that THP caused the problem, can you reverse to this https://github.com/keroro824/HashingDeepLearning/commit/2d10d46b5f6f1eda5d19f27038a596446fc17cee commit?

From my experience, three things can cause Segfault:
1. data path is incorrect
2. TPH has problems
3. Hashtable/hash function parameters are not  matching

If you believe the Config file has no problem, it has to be THP.

**Comment by dszyfelb** [comment]
I have similar issue: 

```
./runme Config_amz.csv
new Network
new Layer
new Node array
Illegal instruction (core dumped)
```

I reversed to  2d10d46 and then I was able to run the experiment.

My TPH configuration:

```
hugeadm --pool-list
      Size  Minimum  Current  Maximum  Default
   2097152     1000     1000     1000        *
1073741824       20       20       20
```

**Comment by jlopezNEU** [comment]
TL;DR on my Ubuntu 19.10 machine I added:
transparent_hugepage=always hugepagesz=1GB hugepages=10 hugepagesz=2MB hugepages=900
to GRUB_CMDLINE_LINUX_DEFAULT in /etc/default/grub. Then update-grub and reboot. You can verify with cat /sys/kernel/mm/transparent_hugepage/enabled and hugeadm --pool-list.

**Comment by yvdriess** [comment]
Note: mmap returns -1 on error, the current code in Node, Layers, etc. checks for NULL:

https://github.com/keroro824/HashingDeepLearning/blob/b7b3bba9bab1fb0e70de7644daa9e2312f21de82/SLIDE/Layer.h#L58

**Comment by xman** [comment]
@yvdriess  This should be resolved in the pull request #22.

**Comment by One-punch24** [comment]
> TL;DR on my Ubuntu 19.10 machine I added:
> transparent_hugepage=always hugepagesz=1GB hugepages=10 hugepagesz=2MB hugepages=900
> to GRUB_CMDLINE_LINUX_DEFAULT in /etc/default/grub. Then update-grub and reboot. You can verify with cat /sys/kernel/mm/transparent_hugepage/enabled and hugeadm --pool-list.

so sorry to interrupt, but I could not find the grub file, should I create one in the location /etc/default/grub ?

**Comment by jlopezNEU** [comment]
Hi,
Im sorry but I don't know why you are missing this file if you have a recent Ubuntu. Perhaps running the update grub command would restore it.
Best regards,
Jose

________________________________

**Comment by One-punch24** [comment]
Thank you so much for replaying. I am a little bit busy these days for some exams.
&nbsp;Can I consult you later? I was so happy and amazed that you answer me so&nbsp;





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "keroro824/HashingDeepLearning"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年11月10日(星期二) 凌晨0:12
收件人:&nbsp;"keroro824/HashingDeepLearning"<HashingDeepLearning@noreply.github.com&gt;;
抄送:&nbsp;"1482806551"<1482806551@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [keroro824/HashingDeepLearning] Segfault (#6)





 Hi,
 Im sorry but I don't know why you are missing this file if you have a recent Ubuntu. Perhaps running the update grub command would restore it.
 Best regards,
 Jose
 
 ________________________________

**Comment by One-punch24** [comment]
I seemed to have solved this problem but still got segment fault. I really want to consult you later to solve some strange problem.
 
Thank you for your help again! You are a nice person treating my problem so seriously.
 





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "keroro824/HashingDeepLearning"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年11月10日(星期二) 凌晨0:12
收件人:&nbsp;"keroro824/HashingDeepLearning"<HashingDeepLearning@noreply.github.com&gt;;
抄送:&nbsp;"1482806551"<1482806551@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [keroro824/HashingDeepLearning] Segfault (#6)





 Hi,
 Im sorry but I don't know why you are missing this file if you have a recent Ubuntu. Perhaps running the update grub command would restore it.
 Best regards,
 Jose
 
 ________________________________

**Comment by Eslam2011** [comment]
Is code run, I have killed problem  ???

**Comment by Eslam2011** [comment]
@jlopezNEU 
Is code work now or not?
